### PR TITLE
[Example] Conditional GAN in pytorch

### PR DESCRIPTION
**Pull request by xiaowu0162**
This pull request contains an example of tuning conditional GANs in pytorch using nni. The pix2pix framework specialized for image-to-image transformation is selected (https://arxiv.org/abs/1611.07004). Some of the code are adapted from the repository maintained by the authors of pix2pix and cycleGAN (https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix). Currently supports tuning on 5 datasets: 'facades', 'night2day', 'edges2handbags', 'edges2shoes', 'maps'. Users can also add and configure their own datasets.

### CODE REVIEW

**Comment by xiaowu0162** [comment]
Hi, this pr is ready for review. @ultmaster @J-shang

File: examples/trials/pix2pix-pytorch/pix2pix.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
+import os
+import logging
+import time
+import argparse
+from collections import namedtuple
+import numpy as np
+import torch
+import nni
+from nni.utils import merge_parameter
```
Comment (ultmaster) [inline examples/trials/pix2pix-pytorch/pix2pix.py:9]: If this and `test.py` are the only two files that involve NNI related changes. We should throw away the rest of the files, and ask users to clone the original repo. We can import files from the original repo by modifying import path.

File: docs/en_US/TrialExample/Pix2pixExample.rst [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+* ``norm``: normalization type
+* ``init_type``: weight initialization method
+* ``lr``: initial learning rate for adam
+* ``beta1``: momentum term of adam
+* ``lr_policy``: learning rate policy
+* ``gan_mode``: type of GAN objective
+* ``lambda_L1``: weight of L1 loss in the generator objective
+
+
+**Experimental**
```
Comment (ultmaster) [inline docs/en_US/TrialExample/Pix2pixExample.rst:30]: Experiments

File: docs/en_US/TrialExample/Pix2pixExample.rst [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+       "lr":{"_type":"choice","_value":[0.0001, 0.0005, 0.001, 0.005, 0.01, 0.1]},
+       "beta1":{"_type":"uniform","_value":[0, 1]},
+       "lr_policy": {"_type":"choice","_value":["linear", "step", "plateau", "cosine"]},
+       "gan_mode": {"_type":"choice","_value":["vanilla", "lsgan", "wgangp"]} ,
+       "lambda_L1": {"_type":"choice","_value":[1, 5, 10, 100, 250, 500]}
+    }
+
+Implemented code directory: :githublink:`search_space.json <examples/trials/pix2pix-pytorch/search_space.json>`
+
+Note: starting from v2.0, the search space is directly included in the config, please find the example here: :githublink:`config_v2.yml <examples/trials/pix2pix-pytorch/config_v2.yml>`
```
Comment (ultmaster) [inline docs/en_US/TrialExample/Pix2pixExample.rst:66]: We are already in v2.0. Why don't we keep `config_v2.yml` only?

File: docs/en_US/TrialExample/Pix2pixExample.rst [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+Some notes on the implementation:
+
+* The trial code for this example is adapted from the `repository maintained by the authors of Pix2pix and CycleGAN <https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix>`__ .  You can also use your previous code directly, refer to `How to define a trial <Trials.rst>`__ for modifying the code.
+* By default, the code uses the dataset "facades". It also supports the datasets 'night2day', 'edges2handbags', 'edges2shoes', and 'maps'.
+* For 'facades', 200 epochs are enough for the model to converge to a point where the difference between different parameters are salient enough for evaluation. If you are using other datasets, please consider increasing the ``n_epochs`` and ``n_epochs_decay`` parameters by either passing them as arguments when calling ``pix2pix.py`` in the config file (discussed below) or changing the ``pix2pix.py`` directly.
+* In this example, we use L1 loss on the test set as the score to report to NNI. Although L1 is by no means a comprehensive measure of image generation performance, at most times it makes sense for evaluating pix2pix models with similar hyperparameter settings.
```
Comment (ultmaster) [inline docs/en_US/TrialExample/Pix2pixExample.rst:83]: Can you explain "evaluating pix2pix models with similar hyperparameter settings."?

File: docs/en_US/TrialExample/Pix2pixExample.rst [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+* Use ``nni.report_final_result(score)`` to report the final result before the trial ends.
+
+Implemented code directory: :githublink:`main.py <examples/trials/pix2pix-pytorch/pix2pix.py>`
+
+Some notes on the implementation:
+
+* The trial code for this example is adapted from the `repository maintained by the authors of Pix2pix and CycleGAN <https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix>`__ .  You can also use your previous code directly, refer to `How to define a trial <Trials.rst>`__ for modifying the code.
+* By default, the code uses the dataset "facades". It also supports the datasets 'night2day', 'edges2handbags', 'edges2shoes', and 'maps'.
+* For 'facades', 200 epochs are enough for the model to converge to a point where the difference between different parameters are salient enough for evaluation. If you are using other datasets, please consider increasing the ``n_epochs`` and ``n_epochs_decay`` parameters by either passing them as arguments when calling ``pix2pix.py`` in the config file (discussed below) or changing the ``pix2pix.py`` directly.
```
Comment (ultmaster) [inline docs/en_US/TrialExample/Pix2pixExample.rst:82]: Do you mean: 200 epochs for search to tell the differences between hyper-parameters. For final training, we need more epochs?

File: docs/en_US/TrialExample/Pix2pixExample.rst [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+To make it easier to run inference and see the generated images, we also incorporate a simple inference code here: :githublink:`test.py <examples/trials/pix2pix-pytorch/test.py>`
+
+To use the code, run the following command:
+
+.. code-block:: bash
+
+   python3 test.py -c CHECKPOINT -p PARAMETER_CFG -d DATASET_NAME -o OUTPUT_DIR
+
+``CHECKPOINT`` is the directory saving the checkpoints (e.g., the ``checkpoints/`` directory in the trial directory). ``PARAMETER_CFG`` is the ``parameter.cfg`` file generated by NNI recording the hyperparameter settings. This file can be found in the trial directory created by NNI.
```
Comment (ultmaster) [inline docs/en_US/TrialExample/Pix2pixExample.rst:120]: Is performance available? Or is there any insights on which hyper-parameter is better?

File: docs/en_US/TrialExample/Pix2pixExample.rst (hunk 3 above)
Comment (xiaowu0162) [reply docs/en_US/TrialExample/Pix2pixExample.rst:66]: We can do that. I will remove the v1 config.

File: docs/en_US/TrialExample/Pix2pixExample.rst (hunk 4 above)
Comment (xiaowu0162) [reply docs/en_US/TrialExample/Pix2pixExample.rst:83]: That was a typo. It was intended to be "similar architecture". L1 is generally used for evaluating models in pix2pix family with similar architectures on the same task. I will fix that in a later commit.

File: docs/en_US/TrialExample/Pix2pixExample.rst (hunk 5 above)
Comment (xiaowu0162) [reply docs/en_US/TrialExample/Pix2pixExample.rst:82]: Yes, although for the "facades" dataset ~200 epochs is enough for the final training too.

File: docs/en_US/TrialExample/Pix2pixExample.rst (hunk 6 above)
Comment (xiaowu0162) [reply docs/en_US/TrialExample/Pix2pixExample.rst:120]: I do have a couple of experiments. I will add the results in a later commit.

File: examples/trials/pix2pix-pytorch/pix2pix.py (hunk 1 above)
Comment (xiaowu0162) [reply examples/trials/pix2pix-pytorch/pix2pix.py:9]: That should be doable. I will fix that.

**Comment by xiaowu0162** [comment]
Hi, this pr is ready again for review

**Review by ultmaster — APPROVED** [review]

File: examples/trials/pix2pix-pytorch/test.py [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+    _, test_params = get_base_params(params_from_cl['dataset'], params_from_cl['checkpoint'])  
+    test_params.update(params_from_cl)
+
+    with open(test_params['parameter_cfg'], 'r') as f:
+        params_from_nni = json.loads(f.readline().strip())['parameters']
+    test_params = merge_parameter(test_params, params_from_nni)
+
+    os.system('mkdir -p {}/input'.format(params_from_cl['output_dir']))
+    os.system('mkdir -p {}/label'.format(params_from_cl['output_dir']))
+    os.system('mkdir -p {}/output'.format(params_from_cl['output_dir']))
```
Comment (J-shang) [inline examples/trials/pix2pix-pytorch/test.py:136]: use `pathlib` maybe better
```
import pathlib
pathlib.Path('/my/directory').mkdir(parents=True, exist_ok=True)
````

**Review by J-shang — APPROVED** [review]

File: examples/trials/pix2pix-pytorch/test.py (hunk 7 above)
Comment (xiaowu0162) [reply examples/trials/pix2pix-pytorch/test.py]: changed

Comment (xiaowu0162) [reply examples/trials/pix2pix-pytorch/test.py]: changed

**Pull request merged**